easydock/preparation_for_docking.py

```diff
@@ -35,43 +38,246 @@ def mol_from_smi_or_molblock(ligand_string):
+        The alignment is done per ring (for example, three alignments are done for three saturated ring) 
+        
+        By default, the conformers will be aligned to the first conformer
+        before the RMS calculation and, as a side-effect, the second will be left
+        in the aligned state.
+
+    As a side-effect, the conformers will be aligned to the first
+    conformer (i.e. the reference) and will left in the aligned state.
+
+    Arguments:
+        - mol:     the molecule
+        - atomIds: (optional) list of atom ids to use a points for
+                    alingment - defaults to all atoms
+        - prealigned: (optional) by default the conformers are assumed
+                        be unaligned and will therefore be aligned to the
+                        first conformer
+
+    Note that the returned RMS matrix is symmetrical, i.e. it is the
+    lower half of the matrix, e.g. for 5 conformers::
+
+        rmsmatrix = [ a,
+                      b, c,
+                      d, e, f,
+                      g, h, i, j]
+
+    where a is the RMS between conformers 0 and 1, b is the RMS between
+    conformers 0 and 2, etc.
+    This way it can be directly used as distance matrix in e.g. Butina
+    clustering.
 
-    def gen_conf(mole, useRandomCoords, randomSeed):
+    """
+
+    cmat_list = []
+    for atom_id_in_ring in atomIds:
+        # if necessary, align the conformers
+        # Note: the reference conformer is always the first one
+        rmsvals = []
+        confIds = [conf.GetId() for conf in mol.GetConformers()]
+        if not prealigned:
+            if atom_id_in_ring:
+                AlignMolConformers(mol, atomIds=atom_id_in_ring, RMSlist=rmsvals)
+            else:
+                AlignMolConformers(mol, RMSlist=rmsvals)
+        else:  # already prealigned
+            for i in range(1, len(confIds)):
+                rmsvals.append(GetConformerRMSFromAtomIds(mol, confIds[0], confIds[i], atomIds=atom_id_in_ring, prealigned=prealigned))
+        # loop over the conformations (except the reference one)
+        cmat_per_ring = []
+        for i in range(1, len(confIds)):
+            cmat_per_ring.append(rmsvals[i - 1])
+            for j in range(1, i):
+                cmat_per_ring.append(GetConformerRMSFromAtomIds(mol, confIds[i], confIds[j], atomIds=atom_id_in_ring, prealigned=True))
+
+        cmat_list.append(np.array(cmat_per_ring))
+
+    cmat_list_array = np.array(cmat_list)
+
+    return list(np.mean(cmat_list_array, axis=0))
+  
+
+def mol_embedding_3d(mol: Chem.Mol, seed: int=43) -> Chem.Mol:
+
+    def find_saturated_ring(mol: Chem.Mol) -> list[list[int]]:
+        ssr = Chem.GetSymmSSSR(mol)
+        saturated_ring_list = []
+        for ring in ssr:
+            is_atom_saturated_array = [mol.GetAtomWithIdx(atom_id).GetHybridization() == Chem.HybridizationType.SP3 for atom_id in ring]
+            if all(is_atom_saturated_array):
```

Comment:
I think `any` is more appropriate then `all`, because this will skip any saturated ring fused with an aromatic ring. If the conformations of such a ring will be very similar they will be cut by `rms` filter. So, it looks safe to use less strict conditions.

Reply:
I see, yeah in that case, we should make it less stricter